ga4gh/converters.py

```diff
@@ -6,35 +6,236 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import collections
+
+import pysam
+
+import ga4gh.protocol as protocol
+
 
 class AbstractConverter(object):
     """
     Abstract base class for converter classes
     """
-    def __init__(self, request, outputStream):
-        self._request = request
-        self._outputStream = outputStream
+    def __init__(self, httpClient):
+        self._httpClient = httpClient
+
+
+##############################################################################
+# SAM
+##############################################################################
+
+
+class SamException(Exception):
+    """
+    Something that went wrong during converting a SAM file
+    """
 
 
 class SamConverter(AbstractConverter):
     """
     Converts a request to a SAM file
     """
-    def __init__(self, searchReadsRequest, outputStream):
-        super(SamConverter, self).__init__(
-            searchReadsRequest, outputStream)
+    def __init__(self, httpClient, searchReadsRequest, outputFile,
+                 binaryOutput):
+        super(SamConverter, self).__init__(httpClient)
+        self._searchReadsRequest = searchReadsRequest
+        self._outputFile = outputFile
+        self._binaryOutput = binaryOutput
 
     def convert(self):
-        raise NotImplementedError()
+        header = self._getHeader()
+        targetIds = self._getTargetIds(header)
+        # pysam can't write to file streams (except for stdout)
+        # http://pysam.readthedocs.org/en/latest/usage.html#using-streams
+        if self._binaryOutput:
+            flags = "wb"
+        else:
+            flags = "wh"  # h for header
+        if self._outputFile:
+            fileString = self._outputFile
+        else:
+            fileString = "-"
```

Comment:
It's purely a personal style thing, but I like the following pattern for these situations:
```python
fileString = "-"
if self._outputFile is not None:
     fileString = self._outputFile
```
This saves us a line of code. We could also have here
```python
fileString = "-" if self._outputFile is None else self._outputFile
```
which is a bit more concise too. It's a matter of taste and not something I feel strongly about, I just thought I'd bring it up.